Servers/controllers/users.controller.ts

```diff
@@ -0,0 +1,211 @@
+            }
+        }
+
+        return res.status(404).json({ data: `user with id: ${userId} not found` })
+
+    } catch (error: any) {
+        return res.status(500).json({
+            data: "internal server error",
+            errorDetails: error.toString()
+        })
+    }
+}
+
+async function createUser(req: IUserAuthRequest, res: Response): Promise<any> {
+    const userIn = req.body as UserIn
+    const users = await getUsers()
+    const usersCtr = users.length
+
+    for (let u of users) {
+        if (u.email === userIn.email) {
+            return res.status(400).json({ data: `user with email: ${userIn.email} already exists` })
+        }
+    }
+
+    const user: User = {
+        id: usersCtr + 1,
+        created_at: Date.now(),
+        last_login: -1,
+        role_id: 1,
+        name: userIn.name,
+        email: userIn.email,
+        password: await bcrypt.hash(userIn.password, process.env.SALT as string)
+    }
+    users.push(user)
+
+    try {
+        await writeFile("./files/users.json", JSON.stringify(users))
+        return res.status(201).json({ data: getUserOut(user) as UserOut })
+    } catch (error: any) {
+        return res.status(500).json({
+            data: "internal server error",
+            errorDetails: error.toString()
+        })
+    }
+}
+
+async function deleteUser(req: IUserAuthRequest, res: Response): Promise<any> {
+    const userId = parseInt(req.params.id);
+    try {
+        const users = await getUsers()
+
+        const filteredUsers = users.filter(u => {
+            return u.id !== userId
+        });
+
+        if (users.length === filteredUsers.length) {
+            return res.status(400).json({ data: `user with id: ${userId} not found` })
+        } else {
+            await writeFile("./files/users.json", JSON.stringify(filteredUsers))
+            return res.status(204).send()
+        }
+
+    } catch (error: any) {
+        return res.status(500).json({
+            data: "internal server error",
+            errorDetails: error.toString()
+        })
+    }
+}
+
+async function updateUser(req: IUserAuthRequest, res: Response): Promise<any> {
+    const userId = parseInt(req.params.id);
+    try {
+        let users = await getUsers()
+        let body = req.body as UserIn
+        const userIndex = users.map(u => u.id).indexOf(userId)
+        const filteredUser: User = users[userIndex];
+
+        if (filteredUser) {
+            // TODO: write a logic to change update email and password of the user
+            const nonUpdatableFields: string[] = Object.keys(body).filter(b =>
+                ["email", "password", "id", "created_at", "last_login", "role_id"].includes(b))
+
+            if (nonUpdatableFields.length !== 0) {
+                return res.status(400).json({ data: `cannot update ${nonUpdatableFields.join(" and ")} of the user` })
+            } else {
+                const userKeys = Object.getOwnPropertyNames(User)
```

Comment:
@HarshP4585 
This is a very important point regarding how we're writing our code, since we've chosen TypeScript with such features and specification.
Please apply these changes to prevent potential errors. 🔃